File: Pyhton3 Implementations/Stack and queue/Circular_queue.py
#IMPLEMENTATION OF CIRCULAR QUEUE
#REAR - INSERT
#FRONT - DELETE
class circular_queue:
    def __init__(self,size):
        self.size=size
        self.c_queue=[None]*size
        self.front=-1
        self.rear=-1
    def enqueue(self,element):
        if(self.front==-1 and self.rear==-1):
            self.front=0
            self.rear=0
            self.c_queue[self.rear]=element
            print(element,"is added to the circular queue")
        elif((self.rear+1)%self.size==self.front):
            print("circular queue is full!!")
        else:
            self.rear=(self.rear+1)%self.size
            self.c_queue[self.rear]=element
            print(element,"is added to the circular queue")
    def dequeue(self):
        if(self.front==-1 and self.rear==-1):
            print("circular queue is empty!!")
        elif(self.front==self.rear):
            print("the dequeued element is:",self.c_queue[self.front])
            self.front=-1
            self.rear=-1
        else:
            print("the dequeued element is:",self.c_queue[self.front])
            self.front=(self.front+1)%self.size
    def display(self):
        if(self.front==-1 and self.rear==-1):
            print("circular queue is empty!!")
        else:
            print("the circular queue is:")
            i=self.front
            while(i!=self.rear):
                print(self.c_queue[i])
                i=(i+1)%self.size
            print(self.c_queue[self.rear])

File: JAVA Implementations/src/Queue/CircularQueue.java
package Queue;

public class CircularQueue {
    int[] circularQueue;
    private static final int DEFAULT_size = 10;
    int front = 0;
    int rear = 0;
    int size = 0;

    public CircularQueue(){
        this(DEFAULT_size);
    }
    public CircularQueue(int size){
        circularQueue = new int[size];
    }

    public boolean insert(int item) throws Exception{
        if(isFull()){
            throw new Exception("Queue is Full!!");
        }
        circularQueue[rear++] = item;
        rear = rear % circularQueue.length;
        size++;
        return true;
    }
    public int remove() throws Exception{
        if(isEmpty()){
            throw new Exception("Queue is Empty");
        }
        int removedItem = circularQueue[front];
        front = (front+1) % circularQueue.length;
        size--;
        return removedItem;
    }
    public boolean isFull(){
        return size == circularQueue.length;
    }
    public boolean isEmpty(){
        return size == 0;
    }
    public int size(){
        return size;
    }

    public void display() throws Exception{
        if(isEmpty()){
            throw new Exception("Empty Queue");
        }
        int i = front;
        do{
            System.out.print(circularQueue[i] + "->");
            i++;
            i %= circularQueue.length;
        }while (i != rear);
        System.out.println("END");
    }

}
